pkg/kubelet/kubeletconfig/controller.go

```diff
@@ -0,0 +1,350 @@
+
+	// startupTracker persists Kubelet startup records, used for crash-loop detection, to a storage layer
+	startupTracker startups.Tracker
+}
+
+// NewController constructs a new Controller object and returns it. Directory paths must be absolute.
+// If the `initConfigDir` is an empty string, skips trying to load the init config.
+// If the `dynamicConfigDir` is an empty string, skips trying to load checkpoints or download new config,
+// but will still sync the ConfigOK condition if you call StartSync with a non-nil client.
+func NewController(initConfigDir string, dynamicConfigDir string, defaultConfig *componentconfig.KubeletConfiguration) *Controller {
+	fs := utilfs.DefaultFs{}
+
+	var initLoader configfiles.Loader
+	if len(initConfigDir) > 0 {
+		initLoader = configfiles.NewFSLoader(fs, initConfigDir)
+	}
+	dynamicConfig := false
+	if len(dynamicConfigDir) > 0 {
+		dynamicConfig = true
+	}
+
+	// Get the current kubelet version; bad-config and startup-tracking information can be kubelet-version specific,
+	// e.g. a bug that crash loops an old Kubelet under a given config might be fixed in a new Kubelet or vice-versa,
+	// validation might be relaxed in a new Kubelet, etc.
+	// We also don't want a change in a file format to break Kubelet upgrades; this makes sure a new kubelet gets
+	// a fresh dir to put its config health data in.
+	// Note that config checkpoints use the api machinery to store ConfigMaps, and thus get file format versioning for free.
+	kubeletVersion := version.Get().String()
+
+	return &Controller{
+		dynamicConfig: dynamicConfig,
+		defaultConfig: defaultConfig,
+		// channels must have capacity at least 1, since we signal with non-blocking writes
+		pendingConfigSource: make(chan bool, 1),
+		configOK:            status.NewConfigOKCondition(),
+		checkpointStore:     store.NewFsStore(fs, filepath.Join(dynamicConfigDir, checkpointsDir)),
+		badConfigTracker:    badconfig.NewFsTracker(fs, filepath.Join(dynamicConfigDir, badConfigTrackingDir, kubeletVersion)),
+		startupTracker:      startups.NewFsTracker(fs, filepath.Join(dynamicConfigDir, startupTrackingDir, kubeletVersion)),
+		initLoader:          initLoader,
+	}
+}
+
+// Bootstrap attempts to return a valid KubeletConfiguration based on the configuration of the Controller,
+// or returns an error if no valid configuration could be produced. Bootstrap should be called synchronously before StartSync.
+func (cc *Controller) Bootstrap() (*componentconfig.KubeletConfiguration, error) {
+	utillog.Infof("starting controller")
+
+	// ALWAYS validate the local (default and init) configs. This makes incorrectly provisioned nodes an error.
+	// These must be valid because they are the foundational last-known-good configs.
+	utillog.Infof("validating combination of defaults and flags")
+	if err := validation.ValidateKubeletConfiguration(cc.defaultConfig); err != nil {
+		return nil, fmt.Errorf("combination of defaults and flags failed validation, error: %v", err)
```

Comment:
One thing is unclear to me in this pr is if the validation of default and init config failed, should kubelet exit with an error? Or keep running if it can but reporting bad status?